### PR TITLE
Update Python versions in the CI workflow.

### PR DESCRIPTION
**Pull request by ezio-melotti**
Same as:
* https://github.com/python/bedevere/pull/520
* https://github.com/python/blurb_it/pull/316

### CODE REVIEW

**Comment by ezio-melotti** [comment]
3.12 is failing with
```py
==================================== ERRORS ====================================
__________________ ERROR collecting tests/test_backport_pr.py __________________
tests/test_backport_pr.py:12: in <module>
    from miss_islington import backport_pr
miss_islington/backport_pr.py:9: in <module>
    from . import tasks, util
miss_islington/tasks.py:9: in <module>
    from celery import bootsteps
/opt/hostedtoolcache/Python/3.12.0-alpha.2/x64/lib/python3.12/site-packages/celery/bootsteps.py:6: in <module>
    from kombu.common import ignore_errors
/opt/hostedtoolcache/Python/3.12.0-alpha.2/x64/lib/python3.12/site-packages/kombu/common.py:14: in <module>
    from .entity import Exchange, Queue
/opt/hostedtoolcache/Python/3.12.0-alpha.2/x64/lib/python3.12/site-packages/kombu/entity.py:7: in <module>
    from .serialization import prepare_accept_content
/opt/hostedtoolcache/Python/3.12.0-alpha.2/x64/lib/python3.12/site-packages/kombu/serialization.py:440: in <module>
    for ep, args in entrypoints('kombu.serializers'):  # pragma: no cover
/opt/hostedtoolcache/Python/3.12.0-alpha.2/x64/lib/python3.12/site-packages/kombu/utils/compat.py:82: in entrypoints
    for ep in importlib_metadata.entry_points().get(namespace, [])
E   AttributeError: 'EntryPoints' object has no attribute 'get'
```

`entry_points()` used to return a `dict`, but now returns an `EntryPoints` instance, that apparently doesn't have a `.get` method. See:
* https://github.com/python/cpython/pull/24782

cc @jaraco

**Comment by hugovk** [comment]
Suggestions for this PR or another.

We can also bump these to:

* `actions/checkout@v3`
* `actions/cache@v3`
* `codecov/codecov-action@v3`

I've got a [script](https://github.com/hugovk/github-tools/blob/main/upgrade_actions.py) to bump actions (works only on tags not branches, but that covers most), although dependabot can take care of it all later.

We can now also remove the `actions/cache` step and use the newish caching built in to `actions/setup-python`: 

* https://github.com/actions/setup-python#caching-packages-dependencies
* https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#caching-packages

Something like:

```yml
      - uses: actions/setup-python@v4
        with:
          python-version: ${{ matrix.python-version }}
          cache: pip
          cache-dependency-path: pyproject.toml
```

**Comment by jaraco** [comment]
Kombu is relying on interfaces removed in 3.12. Fixed in https://github.com/celery/kombu/pull/1601

**Comment by jaraco** [comment]
By the way, you may want to use [this technique](https://github.com/jaraco/skeleton/blob/9e13598ce4b81c2c964dd555fa407bb3ba4cc607/.github/workflows/main.yml#L40-L45) to avoid having to do two-phase commits for each Python version (see actions/setup-python#508).

**Review by hugovk — APPROVED** [review]

**Comment by ezio-melotti** [comment]
Thanks @jaraco for looking into this!

> By the way, you may want to use [this technique](https://github.com/jaraco/skeleton/blob/9e13598ce4b81c2c964dd555fa407bb3ba4cc607/.github/workflows/main.yml#L40-L45) to avoid having to do two-phase commits for each Python version (see https://github.com/actions/setup-python/issues/508).

I'm not sure I understand what you mean here. AFAICS, you are adding `-dev` to the matrix and then use `python-version: ${{ matrix.python }}${{ matrix.dev }}` -- isn't the result equivalent to `python-version: ["3.9-dev", "3.10-dev", "3.11-dev", "3.12-dev"]`?  Also doesn't the `-dev` suffix only apply to the latest version (i.e. `3.12`)?

With the current PR, once `3.13` is released we will just need to do:
```diff
- python-version: ["3.9", "3.10", "3.11", "3.12-dev"]
+ python-version: ["3.9", "3.10", "3.11", "3.12", "3.13-dev"]
```
So, unless I'm missing something, I don't see a reason to handle the `-dev` separately.

**Comment by jaraco** [comment]
> isn't the result equivalent to `python-version: ["3.9-dev", "3.10-dev", "3.11-dev", "3.12-dev"]`?

Yes. Because `-dev` is meaningless after final is released and required before final is released, so it's just meaningless cruft enforced by the maintainers of the action.

> So, unless I'm missing something, I don't see a reason to handle the `-dev` separately.

Except, within about a month of Python 3.12 going final, someone is going to notice that it has `-dev` attached and going to want to remove that cruft, so you'll make another commit or someone will send a PR to remove the `-dev`, which you'll probably accept. Then, when 3.13 comes around a few weeks later, you'll need to add it and have to remember to add `-dev`. If you're lucky, you might get away with just one commit, but in my experience, that approach doesn't generalize.

If instead you apply `-dev` to all versions, then you end up with (a) a consistent way to declare all supported Python versions and (b) the ability to test on pre-release Python versions. Of course, adding `-dev` explicitly to all Python versions makes the matrix noisy and also crufty, but by applying the dev as a separate matrix element, it gets applied mostly silently.

You're welcome to continue to manage it manually - I just wanted to share that I've found a better way that involves less toil.

**Comment by ezio-melotti** [comment]
I now updated the PR to include your suggestion.

**Comment by ezio-melotti** [comment]
I update the branch and 3.12 is now failing.  For more info see:
* python/blurb_it#331
* python/bedevere#542

and the related upstream issues:
* [ ] aiohttp https://github.com/aio-libs/aiohttp/issues/7229
* [ ] frozenlist https://github.com/aio-libs/frozenlist/issues/433
* [ ] yarl https://github.com/aio-libs/yarl/issues/829

**Comment by Mariatta** [comment]
Can we just update to 3.11 on this PR? the 3.12 is still failing.

**Comment by hugovk** [comment]
Yeah, let's split 3.12 into another PR, those other things will take some time, hopefully they will be sorted during the beta. 🤞 

Some other updates we can add to this PR:

```diff
diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index 75a8d93..2e614cb 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -18,8 +18,8 @@ jobs:
         dev: ["-dev"]
 
     steps:
-      - uses: actions/checkout@v2
-      - uses: actions/cache@v1
+      - uses: actions/checkout@v3
+      - uses: actions/cache@v3
         with:
           path: ~/.cache/pip
           key: ${{ runner.os }}-pip-${{ hashFiles('pyproject.toml') }}
@@ -30,7 +30,7 @@ jobs:
           python-version: ${{ matrix.python-version }}${{ matrix.dev }}
       - run: python3 -m pip install -U -r dev-requirements.txt
       - run: pytest --cov=. --cov-report=xml
-      - uses: codecov/codecov-action@v2
+      - uses: codecov/codecov-action@v3
         if: always()
         with:
           token: ${{ secrets.CODECOV_TOKEN }}
```

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -13,7 +13,9 @@ jobs:
 
     strategy:
       matrix:
-        python-version: ["3.9", "3.10"]
+        python-version: ["3.9", "3.10", "3.11", "3.12"]
+        # Workaround for actions/setup-python#508
+        dev: ["-dev"]
```
Comment (hugovk) [inline .github/workflows/ci.yml:18]: https://github.com/actions/setup-python/pull/414 has now been released as https://github.com/actions/setup-python/releases/tag/v4.6.0 and https://github.com/actions/setup-python/releases/tag/v4 so we can do this:

```suggestion
        python-version: ["3.9", "3.10", "3.11", "3.12"]
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -23,9 +25,9 @@ jobs:
           key: ${{ runner.os }}-pip-${{ hashFiles('pyproject.toml') }}
           restore-keys: |
             ${{ runner.os }}-pip-
-      - uses: actions/setup-python@v2
+      - uses: actions/setup-python@v4
         with:
-          python-version: ${{ matrix.python-version }}
+          python-version: ${{ matrix.python-version }}${{ matrix.dev }}
```
Comment (hugovk) [inline .github/workflows/ci.yml:30]: And this:
```suggestion
          python-version: ${{ matrix.python-version }}
          allow-prereleases: true
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -13,7 +13,7 @@ jobs:
 
     strategy:
       matrix:
-        python-version: ["3.9", "3.10"]
+        python-version: ["3.9", "3.10", "3.11", "3.12"]
```
Comment (ezio-melotti) [inline .github/workflows/ci.yml:16]:
```suggestion
        python-version: ["3.9", "3.10", "3.11"]  # see #590 for "3.12"
```
We should comment this out and get the PR merged, and once the 3.12 issues are fixed we can add it back.

**Pull request merged**